src/sentry/processing/realtime_metrics/base.py

```diff
@@ -0,0 +1,38 @@
+import datetime
+
+from sentry.utils import redis
+
+
+class RealtimeMetricsStore:
+    def __init__(
+        self,
+        cluster: redis._RedisCluster,
+        counter_bucket_size: int,
+        counter_ttl: datetime.timedelta,
+        prefix: str = "symbolicate_event_low_priority",
+    ) -> None:
+        if counter_bucket_size <= 0:
+            raise ValueError("bucket size must be at least 1")
+
+        self._counter_bucket_size = counter_bucket_size
+        self.inner = cluster
+        self._counter_ttl: int = int(counter_ttl / datetime.timedelta(milliseconds=1))
+        self._prefix = prefix
+
+    def increment_project_event_counter(self, project_id: int, timestamp: float) -> None:
+        """Increment the event counter for the given project_id.
+
+        The key is computed from the project_id and the timestamp of the symbolication request, rounded
+        down to this object's bucket size. If the key is not currently set to expire, it will be set to expire
+        in ttl seconds.
+        """
+        timestamp = int(timestamp)
+        if self._counter_bucket_size > 1:
+            timestamp -= timestamp % self._counter_bucket_size
```

Comment:
If the bucket size changes midway through an existing bucket, the recorded counts may come out a little funny. Is this correct? Additionally, is this expected behaviour? How often are buckets expected to be changed? I'm thinking this may have an impact on calculations to determine whether a project belongs in a bucket or not.

For example, let's say you start with a bucket size of 10 and switch to a bucket size of 5 7 seconds into the bucket:
| bucket size | timestamp | new hits | inserted redis key:value              |
|-------------|:---------:|----------|------------------------------|
|      10     |     0     |     1    |         `..:..:0`: 1         |
|      10     |     1     |     1    |         `..:..:0`: 2         |
|      10     |     2     |     1    |         `..:..:0`: 3         |
|      10     |     3     |     1    |         `..:..:0`: 4         |
|      10     |     4     |     1    |         `..:..:0`: 5         |
|      10     |     5     |     1    |         `..:..:0`: 6         |
|      10     |     6     |     1    |         `..:..:0`: 7         |
|      5      |     7     |     1    | `..:..:0`: 7,<br>`..:..:5`: 1 |
|      5      |     8     |     1    | `..:..:0`: 7,<br>`..:..:5`: 2 |
|      5      |     9     |     1    | `..:..:0`: 7,<br>`..:..:5`: 3 |

The counts for timestamps 5 and 6 are folded into the 0 bucket, which paints a slightly incorrect picture of the project's behaviour if the bucket size at time of capture isn't recorded.

Similarly, if you go from a smaller bucket to a larger bucket, you get the following behaviour:

| bucket size | timestamp | new hits | redis key:value               |
|-------------|:---------:|----------|-------------------------------|
|      5      |     0     |     1    |          `..:..:0`: 1         |
|      5      |     1     |     1    |          `..:..:0`: 2         |
|      5      |     2     |     1    |          `..:..:0`: 3         |
|      5      |     3     |     1    |          `..:..:0`: 4         |
|      5      |     4     |     1    |          `..:..:0`: 5         |
|      5      |     5     |     1    |  `..:..:0`: 5<br>`..:..:5`: 1 |
|      5      |     6     |     1    | `..:..:0`: 5,<br>`..:..:5`: 2 |
|      10     |     7     |     1    | `..:..:0`: 6,<br>`..:..:5`: 2 |
|      10     |     8     |     1    | `..:..:0`: 7,<br>`..:..:5`: 2 |
|      10     |     9     |     1    | `..:..:0`: 8,<br>`..:..:5`: 2 |

This time, everything from 7-9s is folded into the 0 bucket.

Reply:
Bucket size is, to my understanding, definitely not supposed to change during the lifetime of the `RealtimeMetricsStore`.

Reply:
if you're so inclined you can put the bucket size into the keyname such that metrics data is automatically wiped when you change it. or you remember to change prefix everytime you do

Reply:
This is a great suggestion!  let's put the bucket size in the key, that way we handle it nicely.  when scanning for buckets you use the currently configured bucket size and ignore any keys generated by previous keys.

Proposed new key format: `{prefix}:{bucket_size as int}:{project_id as int}:{timestamp as int}`